### PR TITLE
[SDL 0292] implement VirtualDisplayEncoder improvement for stable frame rate

### LINKED ISSUE #1361: [SDL 0292] Improve VirtualDisplayEncoder for stable frame rate
Proposal: [Improve VirtualDisplayEncoder for stable frame rate](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0292-improve-VDE-for-stable-frame-rate.md)

> This proposal improves the quality of video projection when using Android `VirtualDisplay` and `MediaEncoder`, as Android `MediaEncoder` does not produce a consistent frame rate for rendering `VirtualDisplay`'s surface.

>[SDL-0274](0274-add-preferred-FPS.md) introduced `preferredFPS` parameter specified in `VideoStreamingCapability`. This proposal helps to implement the `preferredFPS` support in the sdl_java_suite platform.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/969

Steering Committee Decision:
>The Steering Committee acknowledged the risks of using the Grafika library as a dependency and that their current implementations are not negatively impacted by this feature, and voted to accept the proposal.

### PR DESCRIPTION
**Pull request by shiniwat**
Fixes #1361 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android, (but not with Java SE, and Java EE)

#### Unit Tests
This PR includes an unit test for Grafika part

#### Core Tests
I have run my SDL VPM app to make sure VPM runs with consistent frame rate.

Core version / branch / commit hash / module tested against: Core 6.1 + preferredFPS integrated (see sdl_core PR at https://github.com/smartdevicelink/sdl_core/pull/3437)
HMI name / version / branch / commit hash / module tested against: our own HMI with VPM capability

### Summary
This implements [SDL 0292] Improve VirtualDisplayEncoder for stable frame rate.

##### Breaking Changes
* N/A

##### Enhancements
* This PR improve the VPM quality.

##### Bug Fixes
* issue #1361 

### Tasks Remaining:
- N/A

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by RHenigan** [comment]
Hello @shiniwat would it be possible to share your VPM app to test this PR?

**Comment by RHenigan** [comment]
Hello @shiniwat during testing how were you able to confirm that the video stream ran at a consistent frame rate?

**Comment by shiniwat** [comment]
Hi @RHenigan,
I have confirmed the frame rate by paying attention to how often onOutputBufferAvailable callback gets called by MediaCodec. It's been done by adding some custom debug message in Proxy.
Because you asked for test VPM app, I am preparing the test app. I'll let you know the app later.

**Comment by shiniwat** [comment]
Hi @RHenigan, my apologies for slow update.
The test app for your request is available at https://github.com/shiniwat/HelloMap. Please let me know if you run into any issues for running the test app.

File: android/sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java
Comment (RHenigan) [inline android/sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java:130]: This would be considered a breaking change, can you please keep the old method and deprecate it and then add in your new method

File: base/src/main/java/com/smartdevicelink/streaming/video/VideoStreamingParameters.java
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/streaming/video/VideoStreamingParameters.java:83]: This would be considered a breaking change, can you please keep the old constructor and deprecate it and then add in your new constructor.

in the old constructor you can default the stableFrameRate value to be true

File: android/sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java
Comment (shiniwat) [reply android/sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java:135]: Thanks for the suggestion. I have restored the previous setStreamingParams method at commit #a4f2539.

File: base/src/main/java/com/smartdevicelink/streaming/video/VideoStreamingParameters.java
Comment (shiniwat) [reply base/src/main/java/com/smartdevicelink/streaming/video/VideoStreamingParameters.java:105]: I have restored previous constructor as you suggested.

**Comment by RHenigan** [comment]
Hello @shiniwat Can you resolve the merge conflicts, some of the code cleanup going on as part of the 5.0 release is creating some minor conflicts.
`VirtualDisplayEncoder` and `VideoStreamingParameters` seem to have conflicts as a result of this code cleanup
`VideoStreamingCapability` however has been impacted by changes to make the setters chainable (please see issue #1421)

While testing I have been unable to validate the video is running at a consistent frame rate. Thank you for sharing the test app but I have trouble with it crashing when the stream starts.

**Comment by shiniwat** [comment]
Hi @RHenigan,
I have merged the current develop branch into my PR branch, so that conflict should have been fixed.
I have updated my HelloMap app as well, and confirmed that VPM works with our IVI system. Please let me know if you still run into crashes. Thanks.

**Review by RHenigan — CHANGES_REQUESTED** [review]

File: base/src/main/java/com/smartdevicelink/proxy/rpc/VideoStreamingCapability.java
Comment (shiniwat) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/VideoStreamingCapability.java:164]: Hi @RHenigan,
Thanks for pointing that out. I have updated the setPreferredFPS accordingly.

**Comment by RHenigan** [comment]
Hello @shiniwat 
Thanks again for your work on this PR! As we've identified SDL 0274 as a dependency for this feature, we will not be able to merge this PR for SDL 0292 until code contributions for SDL 0274 are also reviewed and merged. We've noticed that you've completed some development for SDL 0274 already, so would suggest raising the feature to the Steering Committee as priority for the SDL 2021 Roadmap.

**Comment by shiniwat** [comment]
Hi @RHenigan,
Thanks for your advise. I thought SDL 0274 was planned for SDL Core 7.0 before, but as I checked the project state at https://github.com/smartdevicelink/sdl_core/projects/17, it is not included in Core 7.0 scope. I will make sure it gets prioritized in 2021 roadmap.

**Comment by kostyaBoss** [comment]
@shiniwat Hi! There was found a corner case (by @mrapitis) that if the user starts the video streaming with custom `VideoStreamingParameters` the `getCapability(VIDEO_STREAMING)` is ignored. I've made the suggestion to your code, you can see it [here](https://github.com/shiniwat/sdl_android-1/pull/2). Please, review the suggestions and let me know if there should be some additions

**Comment by shiniwat** [comment]
Hi @kostyaBoss, thanks for the suggestion.
I have reviewed your suggested changes, and have a couple of concerns:
1) You introduced an additional parameter to VideoStreamManager#startRemoteStream, which is basically a breaking change for existing apps.
2) You revealed the potential problematic case where VideoStreamingParameters is explicitly specified. In that case, as you mentioned, VideoStreamingCapability was ignored, but you suggest that VideoStreamingCapability should be respected and should be merged with given parameters. That sounds make sense to me, but probably needs some review from others.
3) It's a little bit unclear to me the difference between CapabilitiesMergeType.SOFT and CapabilitiesMergeType.HARD. Rather than letting developer choose the merge type, I think we can basically take the SOFT merge option for all cases.

**Comment by kostyaBoss** [comment]
@shiniwat Hi! Thanks for your reply. The points 1-3 seem reasonable for us, could you please give more feedback on the point 2 issue when you have some? Thanks

**Comment by joeygrover** [comment]
@shiniwat I would advise against merging any of the suggested changes from that PR as they stand, these changes were not present in the evolution proposal and would need further review before being accepted and merged. This may cause undue work onto you if merged.

@kostyaBoss please note that the changes suggested create breaking changes as well as introducing public API changes, these types of changes must have proper review and follow the defined processes.

**Comment by shiniwat** [comment]
Hi @joeygrover,
Understood your suggestion. As you mentioned, we need another evolution process (i.e. revising the proposal, and then evolution review process) if we take the suggested change.

@kostyaBoss, please let me know the importance of your proposed changes. Sounds like merging VideoStreamingCapability is kind of edge case, right?

**Comment by kostyaBoss** [comment]
@shiniwat I don't think that this is considered to be an edge case, but let me share the reproduction steps there:

- Set up on the HMI some custom parameters (scale 2 for e.x)
- Hardcode the `videoStreaming` parameters on the mobile side
- Call `startRemoteDisplay` with previously hardcoded params
- Observe the behaviour

AR
- The scale will be ignored same as all other parameters from the HMI

ER
- HMI params should not be ignored

I think that's the major scenario

**Comment by shiniwat** [comment]
Hi @kostyaBoss, Understood your point. Because we need to take care of not only preferredFPS, but also other VideoStreamingParameters to cope with VideoStreamingCapability, I believe that issue should be addressed in another proposal rather than revising this one (SDL 0292).
I also need to confirm this issue happens on iOS as well as on Android, right?

**Comment by kostyaBoss** [comment]
@shiniwat Hi! I didn't check the IOS side but should work in the same way

**Comment by shiniwat** [comment]
I double-checked my PR at https://github.com/smartdevicelink/sdl_ios/pull/1613, and just recalled the fact where there's no such VideoStreamingParameters in sdl_ios. Therefore, we don't have to worry about collisions between VideoStreamingParameters and VideoStreamingCapabilities in iOS, i.e. the issue should be specific to sdl_java_suite.
I am thinking about submitting a new proposal that addresses sdl_java_suite specific issue.

**Comment by shiniwat** [comment]
Sorry for not updating previous comment, but the issue raised by @kostyaBoss is discussed in another proposal at [SDL 0323](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0323-align-VideoStreamingParameter-with-capability.md).
As you see in the proposal, we should take lower value approach for maxBitrate and preferredFPS. Current sdl_java_suite already takes care of aligning other parameters, e.g. scale and preferred resolutions are taken from capability, and we should NOT modify other parameters.
Please provide any comment to currently in-review sdl_evolution at https://github.com/smartdevicelink/sdl_evolution/issues/1094. Thanks.

**Comment by shiniwat** [comment]
Because SDL 0323 (Aligning VideoStreamingParameters with VideoStreamingCapability) implementation heavily depends on this PR, I have added another commit (https://github.com/smartdevicelink/sdl_java_suite/pull/1389/commits/3adc0eaf12c453ad4ca4bf2adcff92408bf4350d), which covers implementing SDL 0323.

**Comment by joeljfischer** [comment]
Hi @shiniwat, unfortunately, this PR cannot be reviewed until SDL-0274 (#1268) is implemented, but 0274 is dependent on 0323. Therefore, this PR still cannot be reviewed. Please revert your commit adding changes for 0323 and provide those in a separate PR for review. Thank you.

**Comment by shiniwat** [comment]
Understood. I have reverted the commit for SDL 0323 (3adc0eaf12c453ad4ca4bf2adcff92408bf4350d). I will submit yet another PR to address SDL 0323.

**Comment by shiniwat** [comment]
Regarding SDL 0323, I have submitted another PR at https://github.com/smartdevicelink/sdl_java_suite/pull/1576. Thanks.

**Comment by RHenigan** [comment]
Hello @shiniwat,
I am in the process of reviewing this pull request along with PR #1576 
Because these PRs are dependant on each other I have created an integration branch to be able to test these PRs together before they are merged into develop.
Can you update this PR to point to the new branch [integration/stable_frame_rate](https://github.com/smartdevicelink/sdl_java_suite/tree/integration/stable_frame_rate) instead of develop?

**Comment by shiniwat** [comment]
Hi @RHenigan, according to your request, I have changed the base branch into integration/stable_frame_rate.
BTW, I have noticed that codecov/patch complains some lack of tests (for one of grafika classes). I am going to fix that failure. Thanks.

**Comment by RHenigan** [comment]
Hello @shiniwat, I do not believe we need to add unit tests for the added grafika classes and I will continue my review of this PR

**Review by RHenigan — APPROVED** [review]

**Pull request merged**
